onnxruntime/python/tools/quantization/calibrate.py

```diff
@@ -0,0 +1,232 @@
+    if calib_mode == 'naive':
+        pairs = [tuple([float(min(clean_merged_dict[added_node_output_names[i]])),
+                float(max(clean_merged_dict[added_node_output_names[i+1]]))])
+                for i in range(0, len(added_node_output_names), 2)]
+    elif calib_mode == 'smooth':
+        # Calls smoooth averaging script (number of bootstraps and confidence threshold are adjustable)
+        pairs = [tuple([float(smooth_average(sorted(clean_merged_dict[added_node_output_names[i]]))),
+                float(smooth_average(sorted(clean_merged_dict[added_node_output_names[i+1]])))])
+                for i in range(0, len(added_node_output_names), 2)]
+    final_dict = dict(zip(node_names, pairs))
+    return final_dict
+
+def calculate_scale_zeropoint(node, next_node, rmin, rmax):
+    zp_and_scale = []
+    # adjust rmin and rmax such that 0 is included in the range. This is required
+    # to make sure zero can be uniquely represented.
+    rmin = min(rmin, 0)
+    rmax = max(rmax, 0)
+
+    # We update the output range min and max when next node is clip or relu
+    # With this technique we can remove these 2 ops and
+    # reduce the output range which in turn helps to improve accuracy
+    if next_node.op_type == 'Clip':
+        clip_min = next_node.attribute[0].f
+        clip_max = next_node.attribute[1].f
+        if rmin < clip_min:
+            rmin = clip_min
+        if rmax > clip_max:
+            rmax = clip_max
+    if next_node.op_type == 'Relu':
+        if rmin < 0:
+            rmin = 0
+
+    scale = np.float32((rmax - rmin)/255 if rmin != rmax else 1)
+    initial_zero_point = (0 - rmin) / scale
+    zero_point = np.uint8(round(max(0, min(255, initial_zero_point))))
+
+    zp_and_scale.append(zero_point)
+    zp_and_scale.append(scale)
+    return zp_and_scale
+
+def calculate_quantization_params(model, nbits=8, quantization_thresholds=None):
+    '''
+        Given a model and quantization thresholds, calculates the quantization params.
+    :param model: ModelProto to quantize
+    :param asymmetric_input_types:
+        True: Weights are quantized into signed integers and inputs/activations into unsigned integers.
+        False: Weights and inputs/activations are quantized into unsigned integers.
+    :param output_quantization_thresholds:
+        Dictionary specifying the min and max values for outputs of conv and matmul nodes.
+        The quantization_thresholds should be specified in the following format:
+            {
+                "param_name": [min, max]
+            }
+        example:
+            {
+                'Conv_3:0': [np.float32(0), np.float32(0.5)],
+                'Conv_4:0': [np.float32(1), np.float32(3.5)]
+            }
+    :return: Dictionary containing the zero point and scale values for outputs of conv and matmul nodes.
+        The dictionary format is
+            {
+                "param_name": [min, max]
+            }
+    '''
+    if nbits != 8:
```

Comment:
Remove nbits from function parameter. I think the tool assumes 8 bit, and nbits is not used.